src/test/java/org/jenkinsci/plugins/workflow/graphanalysis/FlowScannerTest.java

```diff
@@ -0,0 +1,492 @@
+        Predicate<FlowNode> outputPredicate = new Predicate<FlowNode>() {
+            @Override
+            public boolean apply(FlowNode input) {
+                if (input instanceof StepAtomNode) {
+                    StepAtomNode san = (StepAtomNode)input;
+                    StepDescriptor sd = san.getDescriptor();
+                    return sd != null && descriptorId.equals(sd.getId());
+                }
+                return false;
+            }
+        };
+        return outputPredicate;
+    }
+
+    Predicate<FlowNode> MATCH_ECHO_STEP = predicateMatchStepDescriptor("org.jenkinsci.plugins.workflow.steps.EchoStep");
+
+    static final class CollectingVisitor implements FlowNodeVisitor {
+        ArrayList<FlowNode> visited = new ArrayList<FlowNode>();
+
+        @Override
+        public boolean visit(@Nonnull FlowNode f) {
+            visited.add(f);
+            return true;
+        }
+
+        public void reset() {
+            this.visited.clear();
+        }
+
+        public ArrayList<FlowNode> getVisited() {
+            return visited;
+        }
+    };
+
+    /** Tests the basic scan algorithm, predicate use, start/stop nodes */
+    @Test
+    public void testSimpleScan() throws Exception {
+        WorkflowJob job = r.jenkins.createProject(WorkflowJob.class, "Convoluted");
+        job.setDefinition(new CpsFlowDefinition(
+                "sleep 2 \n" +
+                "echo 'donothing'\n" +
+                "echo 'doitagain'"
+        ));
+
+        /** Flow structure (ID - type)
+         2 - FlowStartNode
+         3 - SleepStep
+         4 - EchoStep
+         5 - EchoStep
+         6 - FlowEndNode
+         */
+
+        WorkflowRun b = r.assertBuildStatusSuccess(job.scheduleBuild2(0));
+        FlowExecution exec = b.getExecution();
+        AbstractFlowScanner[] scans = {new LinearScanner(),
+                new DepthFirstScanner(),
+                new ForkScanner()
+        };
+
+        List<FlowNode> heads = exec.getCurrentHeads();
+
+        // Iteration tests
+        for (AbstractFlowScanner scan : scans) {
+            System.out.println("Iteration test with scanner: "+scan.getClass());
+            scan.setup(heads, null);
+
+            for (int i=6; i>2; i--) {
+                Assert.assertTrue(scan.hasNext());
+                FlowNode f = scan.next();
+                Assert.assertEquals(Integer.toString(i), f.getId());
+            }
+
+            FlowNode f2 = scan.next();
+            Assert.assertFalse(scan.hasNext());
+            Assert.assertEquals("2", f2.getId());
+        }
+
+        // Block Hopping tests
+        LinearBlockHoppingScanner scanner = new LinearBlockHoppingScanner();
+        Assert.assertFalse("BlockHopping scanner jumps over the flow when started at end", scanner.setup(heads, Collections.EMPTY_SET));
+        List<FlowNode> collectedNodes = scanner.filteredNodes(Collections.singleton(exec.getNode("5")), null, (Predicate)Predicates.alwaysTrue());
```

Comment:
No, the use case I have in mind is something like: _give me all the `milestone` nodes you find starting on a given flow node_ (which is usually the FlowNode of the current step execution). So I guess the predicate should be something like `predicateMatchStepDescriptor(MilestoneStep.class)` and the node the current FlowNode (like [here](https://github.com/jenkinsci/pipeline-milestone-step-plugin/pull/1/files#diff-1918a0bfc343fead0364ffbc8e42b847R90)).